plonk/src/circuit/customized/ultraplonk/lookup_table.rs

```diff
@@ -42,24 +36,38 @@ impl<F: PrimeField> PlonkCircuit<F> {
             self.check_var_bound(table_var.1)?;
         }
         let n = max(lookup_vars.len(), table_vars.len());
-        // update lookup keys for domain separation.
-        let lookup_keys: Vec<Variable> = lookup_vars
-            .iter()
-            .map(|&(key, ..)| self.add_constant(key, &F::from(self.num_table_elems() as u32)))
-            .collect::<Result<Vec<_>, _>>()?;
         let n_gate = self.num_gates();
         (*self.table_gate_ids_mut()).push((n_gate, n));
+        let table_ctr = F::from(self.table_gate_ids_mut().len() as u64);
         for i in 0..n {
-            let (key, val0, val1) = match i < lookup_vars.len() {
-                true => (lookup_keys[i], lookup_vars[i].1, lookup_vars[i].2),
-                false => (self.zero(), self.zero(), self.zero()),
+            let (q_dom_sep, key, val0, val1) = match i < lookup_vars.len() {
+                true => (
+                    table_ctr,
+                    lookup_vars[i].0,
+                    lookup_vars[i].1,
+                    lookup_vars[i].2,
+                ),
+                false => (F::zero(), self.zero(), self.zero(), self.zero()),
             };
-            let (table_val0, table_val1) = match i < table_vars.len() {
-                true => table_vars[i],
-                false => (self.zero(), self.zero()),
+            let (table_dom_sep, table_key, table_val0, table_val1) = match i < table_vars.len() {
+                true => (
+                    table_ctr,
+                    F::from(i as u64),
+                    table_vars[i].0,
+                    table_vars[i].1,
+                ),
+                false => (F::zero(), F::zero(), self.zero(), self.zero()),
             };
             let wire_vars = [key, val0, val1, table_val0, table_val1];
-            self.insert_gate(&wire_vars, Box::new(LookupGate))?;
+
+            self.insert_gate(
+                &wire_vars,
+                Box::new(LookupGate {
+                    q_dom_sep,
+                    table_dom_sep,
+                    table_key,
+                }),
+            )?;
```

Comment:
I'm not very convinced that we need `table_key` for each `Gate`.

Our `fn create_table_and_lookup_variables()` is actually not as general as the name suggests, there's a hidden/implicit assumption which is the online lookup table has a consecutive `table_key` field (namely the truely general API should accept query and lookup tuple of the same length:
```rust
        query_vars: &[(Variable, Variable, Variable)],
        table_vars: &[(Variable, Variable, Variable)],
```
)

If I understand correctly, the main reason for us to have this during arithmetization, it's easier to merge tables if we can just read from this "virtual selector column". My point being, even without this `circuit.table_key_vec()`, we can still calculate this with `table_gate_ids: Vec<(GateId, usize)>` and the implicit assumption that these `table_key` are incremental inside each sub-table. 

^^ am I correct? @chancharles92 
(I'm not suggesting we should remove it, cuz it doesn't burden the final proof size, just increase the memory usage by forcing each gate to have this "virtual selector column". )